Add `f16` and `f128` const eval for binary and unary operationations

tgross35 opened the pull request:
Add const evaluation and Miri support for f16 and f128, including unary and binary operations. Casts are not yet included.

Fixes https://github.com/rust-lang/rust/issues/124583

r? @RalfJung

tgross35 committed:
Enable const evaluation for `f16` and `f128`

This excludes casting, which needs more tests.

tgross35 committed:
Make the unary operator `FloatTy` check exhaustive

Add a spot that was missed in
<https://github.com/rust-lang/rust/pull/121997>.

tgross35 committed:
Add `f16` and `f128` support to Miri

tgross35 commented:
src/tools/miri/tests/pass/float.rs is _really_ tedious to update and I'm not entirely clear what a lot of the magic numbers are intended to test. @RalfJung do you have any suggestions about how this could be cleaned up or made generic to some extent? Compiler-builtins uses a bit of fuzzing + property testing to compare the builtins function to the system function or apfloat https://github.com/rust-lang/compiler-builtins/blob/c04eb9e1afb72bdf943f5e5d77b3812f40526602/testcrate/src/lib.rs#L255, but we don't really have anything to compare it to here.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-17`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9505638721/job/26200997692) [(plain)](https://github.com/rust-lang/rust/commit/0cedee1cf83575d931817c2f78ef1dfcdc66325f/checks/26200997692/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#16 exporting to docker image format
#16 sending tarball 27.6s done
#16 DONE 30.4s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-17]
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-17', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-17/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---
##[group]Testing stage2 compiletest suite=crashes mode=crashes (x86_64-unknown-linux-gnu)

running 227 tests
........................................................................................  88/227
..........................................................................2024-06-13T19:47:04.718766Z ERROR compiletest::runtest: fatal error, panic: "test no longer crashes/triggers ICE! Please give it a mearningful name, add a doc-comment to the start of the test explaining why it exists and move it to tests/ui or wherever you see fit."
[crashes] tests/crashes/124583.rs ... F
............. 176/227
...................................................


failures:

---- [crashes] tests/crashes/124583.rs stdout ----

error: test no longer crashes/triggers ICE! Please give it a mearningful name, add a doc-comment to the start of the test explaining why it exists and move it to tests/ui or wherever you see fit.
thread '[crashes] tests/crashes/124583.rs' panicked at src/tools/compiletest/src/runtest.rs:377:18:
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace


failures:
```

</details>

RalfJung commented:
The file is kind of grown over time, and I did at some point copy a whole bunch of corner case tests over from the rustc test suite -- most of the cast tests are from that, I think.

But I don't know a good way to make this generic, no.

tgross35 commented:
That makes more sense if they came from the rust suite. I just removed the const eval casting until I can take a closer look at that and hopefully clean up that test a bit. Having only unary and binary operations still saves a ton of the headache that was blocking me elsewhere (ICE on negative numbers 😠)

Ready for a look when you get the chance.

matthiaskrgr commented on tests/crashes/124583.rs:
```diff
@@ -1,5 +0,0 @@
-//@ known-bug: rust-lang/rust#124583
-
-fn main() {
-    let _ = -(-0.0f16);
-}
```
Could you add   Fixes #124583 to pr header so that this is auto closed? thx!

tgross35 replied:
I had it in a commit message and thought it would snag it there, but added just in case :)

matthiaskrgr replied:
ah I didn't check the commits :upside_down_face: should be good enough then :)

rust-log-analyzer commented:
The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9506343568/job/26203335279) [(plain)](https://github.com/rust-lang/rust/commit/6863a54f8456c9d131375c601b880328b45dc804/checks/26203335279/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 25 filtered out; finished in 0.00s

     Running tests/ui.rs (obj/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/deps/ui-ff8e5f7317669711)
## Running ui tests in tests/pass for x86_64-apple-darwin
   Compiler: "MIRI_ENV_VAR_TEST"="0" "MIRI_TEMP"="/tmp/miri-uitest-AmkZyz" "RUST_BACKTRACE"="1" /checkout/obj/build/x86_64-unknown-linux-gnu/stage1/bin/miri "--error-format=json" "--sysroot=/checkout/obj/build/x86_64-unknown-linux-gnu/miri-sysroot" "-Dwarnings" "-Dunused" "-Ainternal_features" "-O" "-Zmir-opt-level=4" "-Cdebug-assertions=yes" "-Zui-testing" "--target" "x86_64-apple-darwin" "--out-dir" OUT_DIR
## Running ui tests in tests/pass-dep for x86_64-apple-darwin
   Compiler: "MIRI_ENV_VAR_TEST"="0" "MIRI_TEMP"="/tmp/miri-uitest-AmkZyz" "RUST_BACKTRACE"="1" /checkout/obj/build/x86_64-unknown-linux-gnu/stage1/bin/miri "--error-format=json" "--sysroot=/checkout/obj/build/x86_64-unknown-linux-gnu/miri-sysroot" "-Dwarnings" "-Dunused" "-Ainternal_features" "-O" "-Zmir-opt-level=4" "-Cdebug-assertions=yes" "-Zui-testing" "--target" "x86_64-apple-darwin" "--out-dir" OUT_DIR
test result: ok. 346 passed; 7 ignored;


test result: ok. 24 passed; 8 ignored;
test result: ok. 24 passed; 8 ignored;

## Running ui tests in tests/panic for x86_64-apple-darwin
   Compiler: "MIRI_ENV_VAR_TEST"="0" "MIRI_TEMP"="/tmp/miri-uitest-AmkZyz" "RUST_BACKTRACE"="1" /checkout/obj/build/x86_64-unknown-linux-gnu/stage1/bin/miri "--error-format=json" "--sysroot=/checkout/obj/build/x86_64-unknown-linux-gnu/miri-sysroot" "-Dwarnings" "-Dunused" "-Ainternal_features" "-O" "-Zmir-opt-level=4" "-Cdebug-assertions=yes" "-Zui-testing" "--target" "x86_64-apple-darwin" "--out-dir" OUT_DIR
test result: ok. 15 passed; 1 ignored;

## Running ui tests in tests/fail for x86_64-apple-darwin
## Running ui tests in tests/fail for x86_64-apple-darwin
   Compiler: "MIRI_ENV_VAR_TEST"="0" "MIRI_TEMP"="/tmp/miri-uitest-AmkZyz" "RUST_BACKTRACE"="1" /checkout/obj/build/x86_64-unknown-linux-gnu/stage1/bin/miri "--error-format=json" "--sysroot=/checkout/obj/build/x86_64-unknown-linux-gnu/miri-sysroot" "-Dwarnings" "-Dunused" "-Ainternal_features" "-O" "-Zmir-opt-level=4" "-Cdebug-assertions=yes" "-Zui-testing" "--target" "x86_64-apple-darwin" "--out-dir" OUT_DIR
## Running ui tests in tests/fail-dep for x86_64-apple-darwin
   Compiler: "MIRI_ENV_VAR_TEST"="0" "MIRI_TEMP"="/tmp/miri-uitest-AmkZyz" "RUST_BACKTRACE"="1" /checkout/obj/build/x86_64-unknown-linux-gnu/stage1/bin/miri "--error-format=json" "--sysroot=/checkout/obj/build/x86_64-unknown-linux-gnu/miri-sysroot" "-Dwarnings" "-Dunused" "-Ainternal_features" "-O" "-Zmir-opt-level=4" "-Cdebug-assertions=yes" "-Zui-testing" "--target" "x86_64-apple-darwin" "--out-dir" OUT_DIR
test result: ok.


test result: ok.
test result: ok.

## Running ui tests in tests/native-lib/pass for x86_64-apple-darwin
   Compiler: "MIRI_ENV_VAR_TEST"="0" "MIRI_TEMP"="/tmp/miri-uitest-AmkZyz" "RUST_BACKTRACE"="1" /checkout/obj/build/x86_64-unknown-linux-gnu/stage1/bin/miri "--error-format=json" "--sysroot=/checkout/obj/build/x86_64-unknown-linux-gnu/miri-sysroot" "-Dwarnings" "-Dunused" "-Ainternal_features" "-O" "-Zmir-opt-level=4" "-Cdebug-assertions=yes" "-Zui-testing" "--target" "x86_64-apple-darwin" "-Zmiri-native-lib=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/miri-native-lib/libtestlib.so" "--out-dir" OUT_DIR

test result: ok.


   Compiler: "MIRI_ENV_VAR_TEST"="0" "MIRI_TEMP"="/tmp/miri-uitest-AmkZyz" "RUST_BACKTRACE"="1" /checkout/obj/build/x86_64-unknown-linux-gnu/stage1/bin/miri "--error-format=json" "--sysroot=/checkout/obj/build/x86_64-unknown-linux-gnu/miri-sysroot" "-Dwarnings" "-Dunused" "-Ainternal_features" "-O" "-Zmir-opt-level=4" "-Cdebug-assertions=yes" "-Zui-testing" "--target" "x86_64-apple-darwin" "-Zmiri-native-lib=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/miri-native-lib/libtestlib.so" "--out-dir" OUT_DIR
test result: ok.

 finished in 27.638 seconds
##[endgroup]
---
.......... (120/124)
...F       (124/124)


/checkout/tests/rustdoc-gui/help-page.goml help-page... FAILED
[ERROR] `tests/rustdoc-gui/utils.goml` line 16
    from `tests/rustdoc-gui/help-page.goml` line 23
    from `tests/rustdoc-gui/help-page.goml` line 32: Error: Evaluation failed: The following errors happened: [localStorage item "rustdoc-theme" (of value "ayu") != "null"]: for command `assert-local-storage: {"rustdoc-theme": |theme|}`
Error: ()
Build completed unsuccessfully in 0:03:32
  local time: Thu Jun 13 21:20:07 UTC 2024
  network time: Thu, 13 Jun 2024 21:20:07 GMT
```

</details>

tgross35 commented:
Spurious failure: https://github.com/rust-lang/rust/issues/126430

oli-obk commented on library/std/src/f128/tests.rs:
```diff
@@ -38,3 +38,9 @@ macro_rules! assert_f128_biteq {
         );
     };
 }
+
+#[test]
+fn test_neg_zero() {
+    // only verify const eval negation until we add the rest of the checks
+    let _neg_zero: f16 = -0.0;
+}
```
These are thoroughly covered by the miri tests, I don't think we need it here, but if you wanna keep it, this needs to be f128

RalfJung replied:
Also this isn't even doing any const-eval? If anything it might be hitting some mir-opt codepaths.

oli-obk replied:
yea, const prop was hitting the mir interp code

tgross35 replied:
Ah thanks for the catch, I missed that it was double negation that caused the crash in #124583. Updated `tests/ui/numbers-arithmetic/f16-f128-lit.rs` instead

RalfJung commented on src/tools/miri/tests/pass/float.rs:
```diff
@@ -41,104 +43,43 @@ trait FloatToInt<Int>: Copy {
+float_to_int_fallback!(f16 => i8, u8, i16, u16, i32, u32, i64, u64, i128, u128);
+float_to_int!(f32=> i8, u8, i16, u16, i32, u32, i64, u64, i128, u128);
+float_to_int!(f64 => i8, u8, i16, u16, i32, u32, i64, u64, i128, u128);
+float_to_int_fallback!(f128 => i8, u8, i16, u16, i32, u32, i64, u64, i128, u128);
```
The new impls aren't actually used, are they?

RalfJung commented:
The last commit seems to un-do a bunch of the changes from the earlier commits...?

tgross35 committed:
Remove f16 const eval crash test

Const eval negation was added. This test is now covered by Miri tests,
and merged into an existing UI test.

Fixes <https://github.com/rust-lang/rust/issues/124583>

tgross35 commented:
> The last commit seems to un-do a bunch of the changes from the earlier commits...?

I meant to remove the changes to cast entirely since I don't have a good test (getting tedious as mentioned above), just did a bad job rebasing. History should be cleaner now.

RalfJung commented on src/tools/miri/src/helpers.rs:
```diff
@@ -1201,12 +1201,14 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
                 float_to_int_inner::<Double>(this, src.to_scalar().to_f64()?, cast_to, round),
-            FloatTy::F128 => unimplemented!("f16_f128"),
+            FloatTy::F128 =>
+                float_to_int_inner::<Quad>(this, src.to_scalar().to_f128()?, cast_to, round),
```
This seems to be dead code now that you removed the cast tests?

OTOH, the code is rather obvious, so I am also fine with keeping it.

tgross35 replied:
Yeah you are right. If you have no problem then I think I may as well just leave it, casts won't be far behind.

RalfJung commented:
Maybe update the PR title and description to say what is covered? Mainly unops and binops, it seems.

r=me after that, with or without the helpers.rs change.

tgross35 commented:
I updated the description, feel free to tweak it if I'm missing something. (You or someone else need to r for me).

@rustbot label +F-f16_and_f128

RalfJung commented:
@bors r+

bors commented:
:pushpin: Commit e649042316ec3f832cfe120d04068b6e3e6c1791 has been approved by `RalfJung`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=RalfJung e649042316ec3f832cfe120d04068b6e3e6c1791 -->
<!-- homu: {"type":"Approved","sha":"e649042316ec3f832cfe120d04068b6e3e6c1791","approver":"RalfJung","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
